### PR TITLE
refine-ZK-1148-3

### PR DESCRIPTION
**Pull request by klyvechen**

### CODE REVIEW

File: zktest/src/archive/test2/B85-ZK-3972-3.zul [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<?import org.zkoss.zk.ui.util.*?>
+<!--
+B85-ZK-3972.zul
```
Comment (scribetw) [inline zktest/src/archive/test2/B85-ZK-3972-3.zul]: It doesn't belong to this issue (Again! 👎 )

File: zktest/src/archive/test2/F85-ZK-1148-2.zul [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+Copyright (C) 2018 Potix Corporation. All Rights Reserved.
+
+-->
+
+<zk>
+	<!--<window id="win"-->
+			<!--viewModel="@id('vm')@init('org.zkoss.zktest.test2.F85_ZK_1148_IncludeVM')">-->
+		<!--<button label="show vm attribute"-->
+				<!--onClick='System.out.println("vm = " + win.getAttributes().get(org.zkoss.bind.sys.BinderCtrl.VM));'/>-->
+	<!--</window>-->
```
Comment (scribetw) [inline zktest/src/archive/test2/F85-ZK-1148-2.zul]: Remove these comments if they are not used anymore.

File: zktest/src/archive/test2/F85-ZK-1148.zul [hunk 3]
```diff
@@ -22,4 +22,10 @@
 	<button id="delBtno" label="detach out window" onClick='outWin.detach();'/>
 	<button label="attach window" onClick=''/>
 	<label id="resultLabel"></label>
+	<separator/>
+	<window id="reloadWin"
+		viewModel="@id('vm')@init('org.zkoss.zktest.test2.F85_ZK_1148ReloadVM')">
+		<include id="inc1" src="@load(vm.url) @converter(vm.converter)"/>
+		<button id="reload" label="reload" onClick="@command('reload')"/>
```
Comment (scribetw) [inline zktest/src/archive/test2/F85-ZK-1148.zul]: You added a new button, but no instruction provided. No one else knows how to test and what's the expected result.

File: zkbind/src/org/zkoss/bind/tracker/impl/BindUiLifeCycle.java [hunk 4]
```diff
@@ -205,8 +204,11 @@ public void onEvent(Event event) throws Exception {
 			}
 		});
 		//ZK-1148, add a @destroy annotation method.
-		destroyBinder(comp);
-
+//		destroyBinder(comp);
```
Comment (scribetw) [inline zkbind/src/org/zkoss/bind/tracker/impl/BindUiLifeCycle.java]: Just remove this line

File: zk/src/org/zkoss/zk/ui/impl/PageImpl.java [hunk 5]
```diff
@@ -749,11 +749,10 @@ public void preInit() {
 	}
 
 	public void destroy() {
-		super.destroy();
-
 		for (Component root: this.getRoots()) {
-			root.detach();
+			getDesktop().getWebApp().getConfiguration().invokeDestroy(root);
```
Comment (scribetw) [inline zk/src/org/zkoss/zk/ui/impl/PageImpl.java]: 1. Be aware of NPE of chaining getter
2. Why not just call `afterComponentDetached(root, null);`

File: zktest/src/archive/test2/B85-ZK-3972-3.zul (hunk 1 above)
Comment (klyvechen) [reply zktest/src/archive/test2/B85-ZK-3972-3.zul]: sorry ok

File: zktest/src/archive/test2/F85-ZK-1148-2.zul (hunk 2 above)
Comment (klyvechen) [reply zktest/src/archive/test2/F85-ZK-1148-2.zul]: ok

File: zkbind/src/org/zkoss/bind/impl/BinderImpl.java [hunk 6]
```diff
@@ -293,6 +294,7 @@ private void init() {
 				_log.error("Error when initial phase listener:" + clz, e);
 			}
 		}
+		Executions.getCurrent().getDesktop().getWebApp().getConfiguration().registerCallBack("destroy", new DestroyCallback());
```
Comment (scribetw) [inline zkbind/src/org/zkoss/bind/impl/BinderImpl.java]: Beware of NPE

File: zktest/test/java/org/zkoss/zktest/zats/test2/F85_ZK_1148Test.java [hunk 7]
```diff
@@ -45,11 +45,22 @@ public void testOutWindowDetach() {
 
 	@Test
 	public void testDesktopDestroy() {
+		F85_ZK_1148FileDealer.writeMsg("", false);
 		DesktopAgent da = connect();
 		da.destroy();
-		Assert.assertEquals("DestroyD is called!DestroyC is called!", F85_ZK_1148FileDealer.readMsg());
+		Assert.assertEquals("DestroyD is called!DestroyC is called!IncludeVM is called!", F85_ZK_1148FileDealer.readMsg());
```
Comment (scribetw) [inline zktest/test/java/org/zkoss/zktest/zats/test2/F85_ZK_1148Test.java]: https://github.com/zkoss/zk/pull/1185/files#diff-c52ff1da920c31e56cabddf44f726869R34
You forgot the potential issue

Comment (klyvechen) [reply zktest/test/java/org/zkoss/zktest/zats/test2/F85_ZK_1148Test.java]: ok thanks

File: zkbind/src/org/zkoss/bind/impl/BinderImpl.java (hunk 6 above)
Comment (klyvechen) [reply zkbind/src/org/zkoss/bind/impl/BinderImpl.java]: I put it in the BinderImpl, static{....} , to be sure that registerCallBack only register for once.
The stack is 
	  at org.zkoss.bind.impl.BinderImpl.<clinit>(BinderImpl.java:199)
	  at java.lang.Class.forName0(Class.java:-1)
	  at java.lang.Class.forName(Class.java:348)
	  at org.zkoss.lang.Classes.forNameByThread(Classes.java:274)
	  at org.zkoss.lang.Classes.newInstanceByThread(Classes.java:238)
	  at org.zkoss.bind.BindComposer.newAnnotateBinder(BindComposer.java:429)
	  at org.zkoss.bind.BindComposer.initBinder(BindComposer.java:414)
	  at org.zkoss.bind.BindComposer.doBeforeComposeChildren(BindComposer.java:177)
	  at org.zkoss.zk.ui.impl.UiEngineImpl.doBeforeComposeChildren(UiEngineImpl.java:991)
	  at org.zkoss.zk.ui.impl.UiEngineImpl.execCreateChild0(UiEngineImpl.java:931)
	  at org.zkoss.zk.ui.impl.UiEngineImpl.execCreateChild(UiEngineImpl.java:889)
	  at org.zkoss.zk.ui.impl.UiEngineImpl.execCreate0(UiEngineImpl.java:776)
	  at org.zkoss.zk.ui.impl.UiEngineImpl.execCreate(UiEngineImpl.java:740)
is after starting server, it no need to check the NPE.

File: zkbind/src/org/zkoss/bind/callback/DestroyCallback.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+*/
+package org.zkoss.bind.callback;
+
+import org.zkoss.bind.Binder;
+import org.zkoss.bind.impl.BinderUtil;
+import org.zkoss.zk.ui.AbstractComponent;
+import org.zkoss.zk.ui.Component;
+import org.zkoss.zk.ui.util.Callback;
+
+public class DestroyCallback implements Callback<Component> {
```
[the remark this answers is not in the document]
Comment (klyvechen) [reply zkbind/src/org/zkoss/bind/callback/DestroyCallback.java]: ok

File: zk/src/org/zkoss/zk/ui/util/Configuration.java [hunk 9]
```diff
@@ -1158,6 +1160,26 @@ public void afterComponentDetached(Component comp, Page prevpage) {
...
+		if (callback != null)
+			callback.call(data);
+	}
+
+	/** Register a callback function {@link Callback#call(Object data)}
+	 * @param name the name of the callback function.
+	 * @param callback the callback function to register into Configuration class. 
+	 * @since 8.6.0
+	 */
+	public void registerCallBack(String name, Callback callback) {
```
[the remark this answers is not in the document]
Comment (klyvechen) [reply zk/src/org/zkoss/zk/ui/util/Configuration.java]: ok

File: zk/src/org/zkoss/zk/ui/util/Configuration.java [hunk 10]
```diff
@@ -1158,6 +1160,26 @@ public void afterComponentDetached(Component comp, Page prevpage) {
...
+	public void invokeCallback(String name, Object data) {
+		Callback callback = _callbacks.get(name);
+		if (callback != null)
+			callback.call(data);
+	}
+
+	/** Register a callback function {@link Callback#call(Object data)}
+	 * @param name the name of the callback function.
+	 * @param callback the callback function to register into Configuration class. 
+	 * @since 8.6.0
```
[the remark this answers is not in the document]
Comment (klyvechen) [reply zk/src/org/zkoss/zk/ui/util/Configuration.java]: ok

File: zk/src/org/zkoss/zk/ui/util/Configuration.java [hunk 11]
```diff
@@ -1158,6 +1160,26 @@ public void afterComponentDetached(Component comp, Page prevpage) {
 		}
 	}
 
+	/** Invokes a callback function {@link Callback#call(Object data)}
+	 * @param name the name of the callback function.
+	 * @param data the parameter for the callback function.
+	 * @since 8.6.0
```
[the remark this answers is not in the document]
Comment (klyvechen) [reply zk/src/org/zkoss/zk/ui/util/Configuration.java]: ok

File: zk/src/org/zkoss/zk/ui/util/Configuration.java [hunk 12]
```diff
@@ -127,8 +127,10 @@
 	/** Map(String deviceType, TimeoutURIInfo ti) */
 	private final Map<String, TimeoutURIInfo> _timeoutURIs = Collections
 			.synchronizedMap(new HashMap<String, TimeoutURIInfo>());
-	/** Since 8.0.0 Map(String String) */
+	/** Since 8.0.0 Map(String DataHandlerInfo) */
 	private final Map<String, DataHandlerInfo> _dataHandlers = new HashMap<String, DataHandlerInfo>();
+	/** Since 8.6.0 Map(String Callback) */
```
[the remark this answers is not in the document]
Comment (klyvechen) [reply zk/src/org/zkoss/zk/ui/util/Configuration.java]: ok

File: zk/src/org/zkoss/zk/ui/impl/PageImpl.java [hunk 13]
```diff
@@ -749,11 +749,10 @@ public void preInit() {
 	}
 
 	public void destroy() {
-		super.destroy();
-
 		for (Component root: this.getRoots()) {
-			root.detach();
+			getDesktop().getWebApp().getConfiguration().invokeCallback("destroy", root);
```
[the remark this answers is not in the document]
Comment (klyvechen) [reply zk/src/org/zkoss/zk/ui/impl/PageImpl.java]: for 1, 
The configuration generate webApp, 
the webApp generate desktop, 
the desktop generate page.
here in a page, so may be it no need to worry about the NPE.

[the remark this answers is not in the document]
Comment (klyvechen) [reply zk/src/org/zkoss/zk/ui/impl/PageImpl.java]: 2 ok

File: zkbind/src/org/zkoss/bind/impl/BinderImpl.java [hunk 14]
```diff
@@ -194,6 +195,10 @@ public boolean isDefaultMethod(Method method) {
 		}
 	};
 
+	static {
+		Executions.getCurrent().getDesktop().getWebApp().getConfiguration().registerCallBack("destroy", new DestroyCallback());
```
Comment (scribetw) [inline zkbind/src/org/zkoss/bind/impl/BinderImpl.java]: Looks very strangle to me... No better way?

File: zkbind/src/org/zkoss/bind/tracker/impl/BindUiLifeCycle.java [hunk 15]
```diff
@@ -26,6 +26,7 @@
 import org.zkoss.bind.AnnotateBinder;
 import org.zkoss.bind.BindComposer;
 import org.zkoss.bind.Binder;
+import org.zkoss.bind.callback.DestroyCallback;
```
Comment (scribetw) [inline zkbind/src/org/zkoss/bind/tracker/impl/BindUiLifeCycle.java]: Unused import

Comment (klyvechen) [reply zkbind/src/org/zkoss/bind/tracker/impl/BindUiLifeCycle.java]: ok thanks

**Pull request merged**